vassal-app/src/main/java/VASSAL/build/module/chessclockcontrol/ChessClock.java

```diff
@@ -0,0 +1,826 @@
+        elapsedTime = elapsed;
+        verifiedTime = elapsed; // we assume this is correct
+        clockTicking = ticking;
+        // send it back for verification
+        if (me != null) {
+          GameModule.getGameModule().sendAndLog(new UpdateTimerCommand(me, getName(), elapsedTime, verifiedTime, ticking, false));
+        }
+      }
+
+      setTimerButton();
+      updateTimerColor();
+
+      if (clockTicking || (restore && (elapsedTime > 0))) {
+        ChessClockControl ccc = ChessClockControl.getInstance();
+        if (ccc != null) ccc.showClocks();
+      }
+    }
+
+    @Override
+    protected Command myUndoCommand() {
+      return null;
+    }
+
+    @Override
+    public boolean isLoggable() {
+      return false;
+    }
+  }
+
+  /**
+   * Updates the color of this chess clock's button/display.
+   * Non-ticking clock uses "default" button look.
+   * Ticking clock uses special background and pulsing foreground.
+   */
+  protected void updateTimerColor() {
+    if (!instanceIsActive) return;
+    if (clockTicking) {
+      timerButton.setBackground(tickingBackgroundColor);
+      timerButton.setOpaque(true);
+      timerButton.setBorderPainted(false);
+      timerButton.setForeground(tocking ? tockingFontColor : tickingFontColor);
+    }
+    else {
+      timerButton.setBackground(defaultColor);
+      timerButton.setOpaque(false);
+      timerButton.setBorderPainted(true);
+      timerButton.setForeground(defaultFontColor);
+    }
+  }
+
+  /**
+   * Put timer in initial state, and invisible.
+   */
+  private void initTimerButton() {
+    timerButton.setVisible(false);
+    updateTimerColor();
+  }
+
+  /**
+   * Hides this clock
+   */
+  public void hideClock() {
+    timerButton.setVisible(false);
+  }
+
+  /**
+   * Shows this clock
+   */
+  public void showClock() {
+    timerButton.setVisible(true);
+  }
+
+  /**
+   * Updates the clock display for this clock
+   * (1) Checks new elapsed value from millisecond timer.
+   * (2) Calls setTimerButton to clock time (text) on the button
+   * (3) Updates timer button colors if appropriate
+   */
+  public void updateDisplay() {
+    if (startTime == -1) {
+      startTime = System.currentTimeMillis();
+      return;
+    }
+
+    long currentTime = System.currentTimeMillis();
+    long elapsed = currentTime - startTime;
+    startTime = currentTime;
+
+    if (clockTicking) {
+      elapsedTime += elapsed;
+      if (setTimerButton()) {
+        updateTimerColor();
+      }
+    }
+  }
+
+  /**
+   * @return Clock button text (if any) formatted to be ready to be prepended to time clock string
+   */
+  private String getFormattedButtonText() {
+    return (StringUtils.isEmpty(buttonText) ? "" : buttonText + " ");
+  }
+
+  /**
+   * Updates clock text on the button, based on our configured style
+   * @return half second tick-tock pulse on/off
+   */
+  public boolean setTimerButton() {
+    long time = elapsedTime;
```

Comment:
Can't we use a library here to do these calculations e.g Java 8's Date/Time API?